lib/Parse/ParseExpr.cpp

```diff
@@ -1548,6 +1548,16 @@ ParserResult<Expr> Parser::parseExprPrimary(Diag<> ID, bool isExprBasic) {
     // Objective-C programmers habitually type @"foo", so recover gracefully
     // with a fixit.  If this isn't @"foo", just handle it like an unknown
     // input.
+    if (peekToken().isContextualKeyword("unknown")) {
+      auto UnknownLoc = consumeToken();
+      consumeToken();
+      if (!Tok.is(tok::kw__))
+        goto UnknownCharacter;
+      // Similar to the tok::kw__ case.
+      ExprContext.setCreateSyntax(SyntaxKind::DiscardAssignmentExpr);
+      return makeParserResult(new (Context) DiscardAssignmentExpr(
+          UnknownLoc, consumeToken(), /*Implicit=*/false));
+    }
```

Comment:
This part will be unreachable when "placeholder type" lands. `'@' <identifier> type` is parsed as a `TypeExpr` with an attributed `TypeRepr`. Could you move this logic to `parseExprSequenceElement` handing `tok::at_sign` and construct `UnresolvedPatternExpr` with `AnyPattern` instead of "hacked" `DiscardAssignmentExpr`?